pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java

```diff
@@ -1038,12 +1038,12 @@ private void doSubscribeTopicPartitions(Schema<T> schema,
                 .range(0, numPartitions)
                 .mapToObj(
                     partitionIndex -> {
-                        String partitionName = TopicName.get(topicName).getPartition(partitionIndex).toString();
                         CompletableFuture<Consumer<T>> subFuture = new CompletableFuture<>();
-                        configurationData.setStartPaused(paused);
-                        ConsumerImpl<T> newConsumer = createInternalConsumer(configurationData, partitionName,
-                                partitionIndex, subFuture, createIfDoesNotExist, schema);
+                        String partitionName = TopicName.get(topicName).getPartition(partitionIndex).toString();
                         synchronized (pauseMutex) {
+                            configurationData.setStartPaused(paused);
```

Comment:
Though the current version should synchronize logic, we may need to monitor the performance implication of putting this `createInternalConsumer` under this mutex. As paused is volatile now, I think we may need to consider moving these two lines back to the above `(synchronized (pauseMutex) {)` if the performance is signficantly degraded. I realized `createInternalConsumer`  is a quite expensive op.

Reply:
+1